### DISCUSSION TITLE
Features needed for New Bridges

### DISCUSSION THREAD

**Discussion by nielsmh** [OpenTTD/OpenTTD#9162] [Features]
With PR #9161 currently open, the question for "what would need to go into a New Bridges feature" opens again.

Obvious:
- Transport types allowed (Road, tram, rail, ships, possibly filter on road/railtypes too, like no dirt roads on cantilever steel bridges or maglev on wooden trestles)
- Year of introduction
- Year of obsolescence
- Cost per tile
- Speed limit

Ideas:
- Maybe also tunnels? (They only need an entrance and exit graphics though.)
- Patterns for bridge tiles to use on different lengths of bridges
- Bridge tiles that have pylons and don't have pylons
- Max height of pylons (also relevant for possible #7924 water depth feature)
- Ability to build pylons into water or not
- Clearance required below (number of tile heights underneath the bridge that can't contain other construction, can allow for "deep" ship bridges and better under truss and arch bridges)
- Clearance required above (with bridges over bridges, avoids them clipping into each other)
- Callbacks to select graphics, able to query for neighboring bridges (allow merging bridge decks visually)
- Detailed costs for pylon heights and for individual bridge tiles

**Comment by Andrew350** [comment]
A few of my immediate ideas. Apologies if the PR already addresses some of them, I tried reading through the commits but other than recognizing some keywords, I don't really know what it's doing 😛 Also some of these assume callbacks will be available per tile, which I'm not sure will be possible, but anyway:
 - separate sprites for front and rear parts of bridge heads (absolutely necessary, IMHO)
 - allow querying height above ground (e.g for solid viaducts when low, open arches/pillars when high). Could also allow pillars themselves to change graphics depending on height and/or location, so they can be wider at the bottom, etc.
- allow querying tile_class/groundtype (for snow-aware bridges, gaps over roads, etc.), also querying tile_slope to allow making pillars/bridge pieces look nicer on slopes.
 - maybe allow querying date of construction/age?
 - ability to define the underside as impassible/no construction allowed (possibly per-tile?), to allow for visual tricks like floating bridges, or the aforementioned solid viaducts etc. without having roads/ships/whatnot going right through it. Not sure if this one is possible or not, but would be neat 🙂

> Transport types allowed (Road, tram, rail, ships, possibly filter on road/railtypes too, like no dirt roads on cantilever steel bridges or maglev on wooden trestles)

Possibly in addition to this, a flag to just disable the bridge overlay sprites (maybe not including tramtypes and/or catenary), rather than disallowing the whole road/railtype? It might be contentious with road/tram/railtype set authors, but would for example allow trestles with non-solid (see-through) decks to remain compatible with railtypes which don't normally provide nice-looking overlays for such a bridge.

> Patterns for bridge tiles to use on different lengths of bridges

You may have had this in mind, but also being able to define "arrays" of layouts would be neat. So for example at certain lengths (in this example multiples of 11) a bridge could do [3-part-layout, 5-part-layout, 3-part-layout], but at other lengths just repeat [5-part-layout] or [3-part-layout] or whatever.

Also each layout should be able to be marked as repeatable or not, which would allow for both tile-able and one-off designs, or a mix like above.


There are some other ideas that are probably a bit too crazy (working draw bridges anyone? 😛) so I'll leave that out for now. But everything else here so far is +1000 from me 👍

**Reply by kiwitreekor** [reply]
terrain type and construction date is already available 🙂

**Reply by PeterN** [reply]
Having pillars that "blocked" the tile below if it was empty looked pretty nice back in my original NewGRF bridges work. That was when bridge tiles actually existed though.

What about special handling for adjacent bridges?

**Reply by nielsmh** [reply]
I was thinking that blocking the tile below could be covered by "clearance below" requirements, but maybe not quite. "Tile extends all the way down" is probably a separate flag yes.

Another flag would be "impassable pylons", e.g. for stone arch bridges you generally wouldn't be able to build diagonal rail underneath, since the stone supports would be in the way.

Most of these flags would be per-tile, and actually could add some computational cost to building bridges, since (because bridges only exist as the two bridgeheads) the game would need to search for all bridges passing over a tile, and then figure out the layout of every bridge and determine the permissions the bridge tiles above give.

**Comment by kiwitreekor** [comment]
Some Ideas...
 * New format of layout table can be required, to support custom patterns, etc.
   * Should we separate rail/road/monorail/maglev tables?
 * How can we set "per-tile properties"?
   * Add callback like cb 36? (with variable querying position of tile)
 * Currently newgrf debug tool cannot show variables of middle bridge tiles

**Comment by auge8472** [comment]
Further idea:

- Overwriting the bridges speed limit *in the building menu* from within a road or rail NewGRF. Especially with NRT road types that often defines slow speed limits it looks really odd if the road is limited i.e. to 30kph but the road types bridges says it is limited to i.e. 128kph in the menu (even the speed limit *is* effectively that of the road type).
- Restricting the buildable bridges by their own intended speed limits in a rod or rail NewGRF. Example: as a set author I want to restrict the accessible bridges to slow types, the ones that are typically the "old", first introduced bridge types. Pitfall: this could end with forbidden bridge building depending on the combination of the bridge set and the restricting road type (road forbids access to bridges faster than 70kph but the bridge set does not contain a slower bridge).
- Forbid bridge or tunnel building in a road or rail type similar to being able to forbid houses or rail crossings.

**Comment by JGRennison** [comment]
Some bridge sets add types are only intended to be used in combination, not on their own, e.g. the Modular Wider Bridges Set. With more/richer bridge types, more GRFs may also do this.
It may be useful to add flag(s) which prevents towns or AI/GS from building bridges of this type (this is already implemented in my branch and in the Modular Wider Bridges Set).

On pillars, at the moment in my branch the presence of pillars can be set on a per-corner and per-edge basis for each bridge part, [see here](https://htmlpreview.github.io/?https://github.com/JGRennison/OpenTTD-patches/blob/jgrpp/docs/newgrf-additions.html#bridge_pillar_flags). This is currently just used for stations (which have corresponding flags for which pillars are not allowed), but it would make sense for something like this to also be used for road/track/etc underneath, as suggested above.

**Reply by nielsmh** [reply]
One of my suggestions is actually to allow callback-driven graphics selection for bridges, which would allow for the effects of the Modular Wider Bridges, without the player needing to select different bridge types for each adjacent bridge, and could perhaps also give better rendering of cases where the adjacent bridges are different lengths.

**Comment by Ufiby** [comment]
I'm a little surprised that there is any interest in developing the bridge
I have a suggestion, only 3 suggestions
**1.Necessary**
The cost of the bridge, the size is only 1 byte (up to 255), which is quite very cheap, so I want to redo up to 2 bytes (up to 65535). Because I want a miracle bridge to cost more than a bridge that is more than 40 tiles long
**2.New feature "Hidden"**
This is a useful feature to hide from purchases in the menu. Therefore, it is better to hide the Maglev and the monorail from the wooden bridge. And not only that, but I don't need bridges more than 10 types of bridges maglevs and monorails
**3.New feature "symmetric x4" This is desirable.**
So currently OpenTTD bridges are symmetric x2 (**-0(23)(23)1-**), for cable-stayed and suspension bridges is too short. It would be more interesting to have very long bridges. But only conditionally, by default, they build symmetric bridges x2, if you want x4 - you need to add the flag (x4 is symmetric). Table ID for symmetric bridges x4-up to F(up to 16).
![BRIDGE_TEST](https://user-images.githubusercontent.com/46504018/116823943-20de1e00-ab90-11eb-8fcb-ddb52db2332f.png)

**Reply by Ufiby** [reply]
![BRIDGE_TEST_1](https://user-images.githubusercontent.com/46504018/116824331-3f451900-ab92-11eb-8aed-8a0e4227558d.png)
![BRIDGE_TEST_2](https://user-images.githubusercontent.com/46504018/116824334-42d8a000-ab92-11eb-8550-9697b670107a.png)

**Reply by nielsmh** [reply]
This is what I mean by
> * Patterns for bridge tiles to use on different lengths of bridges

in my opening post :)
